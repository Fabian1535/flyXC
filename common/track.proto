syntax = "proto2";

// A single track.
// Fixes are differential encoded.
message Track {
  optional string pilot = 1;
  repeated sint32 lat = 2 [packed = true];
  repeated sint32 lon = 3 [packed = true];
  repeated sint32 alt = 4 [packed = true];
  repeated uint32 ts = 5 [packed = true];
}

// All the tracks for a single file.
// Stored in the database.
message TrackGroup {
  repeated Track tracks = 2;
}

// Ground altitude for a single Track.
message GroundAltitude {
  repeated sint32 altitudes = 1 [packed = true];
  // True when errors were encountered while retrieveing the altitudes.
  optional bool has_errors = 2;
}

// Ground altitude for a Track group.
// Stored in the database.
message GroundAltitudeGroup {
  repeated GroundAltitude ground_altitudes = 1;
}

message MetaTrackGroup {
  optional int64 id = 1;
  optional uint32 num_postprocess = 2;
  optional bytes track_group_bin = 3;
  // This field is only present when there are some altitude data.
  // The GroundAltitude can still have has_errors = true if not all the data
  // were retrieved correctly.
  optional bytes ground_altitude_group_bin = 4;
}

message MetaTracks {
  repeated bytes meta_track_groups_bin = 1;
}