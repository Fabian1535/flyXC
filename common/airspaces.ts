import { Point } from './track';

type Polygon = Point[];

// Flags of the airspaces.
export const enum Flags {
  AIRSPACE_PROHIBITED = 1 << 0,
  AIRSPACE_RESTRICTED = 1 << 1,
  AIRSPACE_DANGER = 1 << 2,
  AIRSPACE_OTHER = 1 << 3,
  BOTTOM_REF_GND = 1 << 4,
  TOP_REF_GND = 1 << 5,
}

// Returns a unique tile id for a given (x, y, zoom).
export function tileId(zoom: number, point: Point): number {
  return ((1 << zoom) * point.y + point.x) * 32 + zoom;
}

// Returns the color of an airspace given the flags property.
// `alpha` should be in the range 00-ff
export function airspaceColor(flags: number, alpha: number): string {
  const alphaStr = String(alpha).padStart(2, '0');
  if (flags & Flags.AIRSPACE_PROHIBITED) {
    return `#bf4040${alphaStr}`;
  }
  if (flags & Flags.AIRSPACE_RESTRICTED) {
    return `#bfbf40${alphaStr}`;
  }
  if (flags & Flags.AIRSPACE_DANGER) {
    return `#bf8040${alphaStr}`;
  }
  return `#808080${alphaStr}`;
}

// Returns the category of the airspace.
export function airspaceCategory(flags: number): string {
  if (flags & Flags.AIRSPACE_PROHIBITED) {
    return `prohibited`;
  }
  if (flags & Flags.AIRSPACE_RESTRICTED) {
    return `restricted`;
  }
  if (flags & Flags.AIRSPACE_DANGER) {
    return `danger`;
  }
  return `other`;
}

// Returns whether the point is inside the polygon feature.
export function isInFeature(point: Point, feature: any): boolean {
  const ratio = 256 / feature.extent;
  const polygons = classifyRings(feature.loadGeometry());
  for (const rings of polygons) {
    // The point must be in the outer ring.
    let isIn = isInPolygon(point, rings[0], ratio);
    if (isIn) {
      for (let i = 1; i < rings.length; ++i) {
        // The point must not be in any hole.
        isIn = isIn && !isInPolygon(point, rings[i], ratio);
      }
    }
    if (isIn) {
      return true;
    }
  }

  return false;
}

export function getAspTileUrl(x: number, y: number, z: number): string {
  return `https://airspaces.storage.googleapis.com/tiles/20200714/${z}/${x}/${y}.pbf`;
}

// Returns whether the point is in the polygon.
function isInPolygon(point: Point, polygon: Point[], ratio: number): boolean {
  const { x, y } = { x: point.x / ratio, y: point.y / ratio };

  let isIn = false;

  let [xa, ya] = [polygon[0].x, polygon[0].y];
  for (let j = 1; j < polygon.length; j++) {
    const [xb, yb] = [polygon[j].x, polygon[j].y];

    if (ya > y != yb > y && x < ((xb - xa) * (y - ya)) / (yb - ya) + xa) {
      isIn = !isIn;
    }
    [xa, ya] = [xb, yb];
  }

  return isIn;
}

// Code adapted from https://github.com/mapbox/vector-tile-js

// Returns an array of polygons.
// Each polygon is an array ring.
// The first ring in this array is the outer ring. Following rings are holes.
export function classifyRings(rings: Point[][]): Polygon[][] {
  const len = rings.length;

  if (len <= 1) {
    return [rings];
  }

  const polygons: Polygon[][] = [];
  let polygon: Polygon[] | null = null;
  let ccw: boolean | null = null;

  for (let i = 0; i < len; i++) {
    const area = signedArea(rings[i]);
    if (area === 0) {
      continue;
    }

    if (ccw == null) {
      ccw = area < 0;
    }

    if (ccw === area < 0) {
      // Create a new polygon when the winding is the same as the first polygon.
      if (polygon) {
        polygons.push(polygon);
      }
      polygon = [rings[i]];
    } else {
      // Pushes holes in the current polygon.
      polygon?.push(rings[i]);
    }
  }
  if (polygon) {
    polygons.push(polygon);
  }

  return polygons;
}

// Computes the area of a polygon.
// See https://en.wikipedia.org/wiki/Shoelace_formula.
function signedArea(polygon: Polygon): number {
  let sum = 0;
  const len = polygon.length;
  for (let i = 0, j = len - 1; i < len; j = i++) {
    const p1 = polygon[i];
    const p2 = polygon[j];
    sum += (p2.x - p1.x) * (p1.y + p2.y);
  }
  return sum;
}
