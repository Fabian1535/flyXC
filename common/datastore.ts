/* eslint-disable @typescript-eslint/no-var-requires */
const { Datastore } = require('@google-cloud/datastore');
/* eslint-enable @typescript-eslint/no-var-requires */

import zlib from 'zlib';

import { ProtoMetaTrackGroup } from './track';

const datastore = new Datastore();

const key_symbol: unique symbol = datastore.KEY;

// A track as stored in datastore.
// The ArrayBuffer zre stored gzipped in the datastore.
// They are automatically compressed and expanded when using the API in this file.
export interface TrackEntity {
  [key_symbol]?: any;
  created: Date;
  city?: string;
  country?: string;
  // TODO: delete after migration
  data?: ArrayBuffer;
  // GroundAltitudeGroup.
  ground_altitude_group?: ArrayBuffer;
  // Whether there are some post-processing errors
  has_postprocess_errors: boolean;
  hash: string;
  // Incremented after each post processing.
  num_postprocess: number;
  // Simplified path (polyline encoded).
  path?: string;
  // ProtoTrackGroup.
  track_group?: ArrayBuffer;
  // url of the track (when loaded from an url).
  url?: string;
  valid: boolean;
}

// Saves an entity to the datastore.
// Large protobufs are automatically gzipped before being stored.
// Returns the id of the track.
export async function saveTrack(track: TrackEntity): Promise<number> {
  // Use the existing entity when it has a key.
  const key = track[key_symbol] ?? datastore.key('Track');

  await datastore.save({
    key,
    excludeFromIndexes: ['data', 'ground_altitude_group', 'path', 'track_group'],
    data: {
      ...track,
      // TODO: delete after migration
      data: zipOrUndefined(track.data),
      ground_altitude_group: zipOrUndefined(track.ground_altitude_group),
      track_group: zipOrUndefined(track.track_group),
    },
  });
  return key.id;
}

// Retrieves a meta track group given its hash.
//
// Note:
// There can be no select clause here are fields are not indexed.
// See https://cloud.google.com/datastore/docs/concepts/queries#projection_queries
export async function retrieveMetaTrackGroupByHash(hash: string): Promise<ProtoMetaTrackGroup | null> {
  if (process.env.USE_CACHE) {
    const query = datastore.createQuery('Track').filter('hash', hash).limit(1);
    const entities = (await datastore.runQuery(query))[0];
    if (entities.length == 1) {
      return entityToMetaTrackGroup(entities[0]);
    }
  }
  return null;
}

// Retrieves a meta track group given its url.
//
// Note:
// There can be no select clause here are fields are not indexed.
// See https://cloud.google.com/datastore/docs/concepts/queries#projection_queries
export async function retrieveMetaTrackGroupByUrl(url: string): Promise<ProtoMetaTrackGroup | null> {
  if (process.env.USE_CACHE) {
    const query = datastore.createQuery('Track').filter('url', url).filter('valid', true).limit(1);
    const entities = (await datastore.runQuery(query))[0];
    if (entities.length == 1) {
      return entityToMetaTrackGroup(entities[0]);
    }
  }
  return null;
}

// Retrieves a meta track group given its url.
//
// Note:
// There can be no select clause here are fields are not indexed.
// See https://cloud.google.com/datastore/docs/concepts/queries#projection_queries
export async function retrieveMetaTrackGroupsByIds(ids: Array<number | string>): Promise<ProtoMetaTrackGroup[]> {
  if (process.env.USE_CACHE) {
    const entities = await datastore.get(ids.map((id) => datastore.key(['Track', Number(id)])));
    return entities[0].map(entityToMetaTrackGroup);
  }
  return [];
}

// Retrieves a track entity given its datastore id.
// Gzipped fields are expanded in the return value.
export async function retrieveTrackById(id: number | string): Promise<TrackEntity | null> {
  if (process.env.USE_CACHE) {
    const entities = await datastore.get(datastore.key(['Track', Number(id)]));
    if (entities[0] != null) {
      return unzipEntity(entities[0]);
    }
  }
  return null;
}

// Converts a gzipped entity to a Meta Track Group.
function entityToMetaTrackGroup(entity: TrackEntity): ProtoMetaTrackGroup {
  if (!(key_symbol in entity)) {
    throw new Error('An ID is required');
  }

  entity = unzipEntity(entity);

  return {
    id: entity[key_symbol].id as number,
    num_postprocess: entity.num_postprocess,
    track_group_bin: entity.track_group,
    ground_altitude_group_bin: entity.ground_altitude_group,
  };
}

function zipOrUndefined(buffer: ArrayBuffer | undefined): ArrayBuffer | undefined {
  return buffer == null ? undefined : zlib.gzipSync(buffer, { level: 9 });
}

function unzipOrUndefined(buffer: ArrayBuffer | undefined): ArrayBuffer | undefined {
  return buffer == null ? undefined : zlib.gunzipSync(buffer);
}

// Expands entity gzipped fields.
function unzipEntity(track: TrackEntity): TrackEntity {
  return {
    ...track,
    // TODO: delete after migration
    data: unzipOrUndefined(track.data),
    track_group: unzipOrUndefined(track.track_group),
    ground_altitude_group: unzipOrUndefined(track.ground_altitude_group),
  };
}
