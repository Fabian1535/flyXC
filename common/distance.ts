import { getDistance } from 'geolib';

import { RuntimeFixes, RuntimeTrack } from './track';

// Computes the maximum distance between two consecutive fixes of a track.
export function computeMaxDistance(fixes: RuntimeFixes): number {
  let max = 0;
  for (let i = 1; i < fixes.lat.length; i++) {
    const a = { lat: fixes.lat[i - 1], lon: fixes.lon[i - 1] };
    const b = { lat: fixes.lat[i], lon: fixes.lon[i] };
    max = Math.max(getDistance(a, b), max);
  }
  return max;
}

// Finds the closest fix to {lat, lon} across all the tracks.
export function findClosestFix(tracks: RuntimeTrack[], lat: number, lon: number): { track: number; ts: number } | null {
  let track = 0;
  let ts: number | null = null;
  let distance = 10000;
  const ref = { lat, lon };
  tracks.forEach((t, tIdx) => {
    for (let fixIdx = 0; fixIdx < t.fixes.lat.length; ) {
      const lat = t.fixes.lat[fixIdx];
      const lon = t.fixes.lon[fixIdx];
      const d = getDistance(ref, { lat, lon });
      if (d < distance) {
        ts = t.fixes.ts[fixIdx];
        track = tIdx;
        distance = d;
        fixIdx++;
      } else {
        fixIdx += Math.max(1, Math.floor((d - distance) / t.maxDistance));
      }
    }
  });

  return ts ? { track, ts } : null;
}
