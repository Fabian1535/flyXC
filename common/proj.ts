import SphericalMercator from '@mapbox/sphericalmercator';

import { LatLon, Point } from './track';

const TILE_SIZE = 256;
const mercator = new SphericalMercator({ size: TILE_SIZE });

// Returns the coordinates of the tile and the pixel inside the tile.
export function tileCoordinates(latLon: LatLon, zoom: number): { tile: Point; px: Point } {
  const screenPx = mercator.px([latLon.lon, latLon.lat], zoom);

  const tile = {
    x: Math.floor(screenPx[0] / TILE_SIZE),
    y: Math.floor(screenPx[1] / TILE_SIZE),
  };

  const px = {
    x: screenPx[0] % TILE_SIZE,
    y: screenPx[1] % TILE_SIZE,
  };

  return {
    tile,
    px,
  };
}
