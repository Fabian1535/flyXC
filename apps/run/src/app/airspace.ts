import { fetchResponse, Flags, getAspTileUrl, isInFeature, pixelCoordinates, Point, protos } from '@flyxc/common';
import async from 'async';
import { VectorTile } from 'mapbox-vector-tile';
import { LRU, lru } from 'tiny-lru';

// Zoom level for the airspaces tiles.
const ZOOM_LEVEL = 12;

// Look MARGIN_METER above max altitude.
const MARGIN_METER = 200;

// Average airspace tile size.
// The average size is ~250 bytes at level 12 (395MB	for 1649364 tiles).
// The maximum size is ~2.2KB.
const ASLP_LRU_CAPACITY = 10000;

// null is used when there is no tile at the given location.
let aspCache: LRU<ArrayBuffer | null> | null = null;

export function getIndicesByUrlAndPx(
  track: protos.Track,
  maxNumTiles = Number.MAX_SAFE_INTEGER,
): {
  indexesByTileUrl: Map<string, number[]>;
  pxInTiles: Array<Point>;
} {
  const indexesByTileUrl = new Map<string, number[]>();
  const pxInTiles: Array<Point> = [];

  for (let i = 0; i < track.lat.length; i++) {
    const lon = track.lon[i];
    const lat = track.lat[i];
    const { tile, px } = pixelCoordinates({ lat, lon }, ZOOM_LEVEL);
    pxInTiles.push(px);
    const url = getAspTileUrl(tile.x, tile.y, ZOOM_LEVEL);
    if (!indexesByTileUrl.has(url)) {
      indexesByTileUrl.set(url, []);
    }
    indexesByTileUrl.get(url)?.push(i);
    if (indexesByTileUrl.size >= maxNumTiles) {
      break;
    }
  }

  return { indexesByTileUrl, pxInTiles };
}

// Retrieves the airspaces for the given track.
export async function fetchAirspaces(track: protos.Track, altitude: protos.GroundAltitude): Promise<protos.Airspaces> {
  const cache = getAspCache();
  const { indexesByTileUrl, pxInTiles } = getIndicesByUrlAndPx(track, 500);

  // Downloading tiles might return 404 if there is not airspace.
  // We then can not rely on 404 to detect errors.
  await async.eachLimit(indexesByTileUrl.keys(), 5, async (url: string) => {
    let buffer = cache.get(url);
    if (buffer === undefined) {
      try {
        const response = await fetchResponse(url, {
          retry: 3,
          timeoutS: 5,
          retryOnTimeout: true,
        });
        buffer = response.ok ? await response.arrayBuffer() : null;
      } catch (e) {
        buffer = null;
      }
      cache.set(url, buffer);
    }
  });

  // Retrieve fix indexes for every airspaces.
  const gndAlt = altitude.altitudes;
  const indexesByAirspaceKey = new Map<string, number[]>();
  const featureByKey = new Map<string, any>();
  const maxAltitude = Math.max(...track.alt) + MARGIN_METER;
  for (const [url, indexes] of indexesByTileUrl) {
    const buffer = cache.get(url);
    if (buffer == null) {
      continue;
    }
    const aspLayer = new VectorTile(new Uint8Array(buffer)).layers.asp;

    for (let i = 0; i < aspLayer.length; i++) {
      const feature = aspLayer.feature(i);
      const props = feature.properties;
      const key = getAspFeatureKey(feature);
      for (const fixIdx of indexes) {
        const flags = props.flags as number;
        const bottom = (props.bottom as number) + (flags & Flags.FloorRefGnd ? (gndAlt[fixIdx] as number) : 0);
        if (bottom > maxAltitude) {
          continue;
        }
        if (isInFeature(pxInTiles[fixIdx], feature)) {
          if (!indexesByAirspaceKey.has(key)) {
            indexesByAirspaceKey.set(key, []);
            featureByKey.set(key, feature);
          }
          indexesByAirspaceKey.get(key)?.push(fixIdx);
        }
      }
    }
  }

  // Compute index ranges.
  const rangesByAirspaceKey = new Map<string, Array<[number, number]>>();
  for (const [key, indexes] of indexesByAirspaceKey) {
    indexes.sort((a, b) => (a < b ? -1 : 1));
    let start = indexes[0];
    let current = indexes[0];
    rangesByAirspaceKey.set(key, []);
    for (let i = 1; i < indexes.length; i++) {
      const next = indexes[i];
      if (next == current + 1) {
        current = next;
        continue;
      }
      rangesByAirspaceKey.get(key)?.push([start, current]);
      start = next;
      current = next;
    }
    rangesByAirspaceKey.get(key)?.push([start, current]);
  }

  // Check if the tracks crosses the airspaces.
  const intoRangesByAirspaceKey = new Map<string, boolean[]>();
  for (const [key, ranges] of rangesByAirspaceKey) {
    const props = featureByKey.get(key).properties;
    const intoRanges: boolean[] = [];
    intoRangesByAirspaceKey.set(key, intoRanges);
    for (const [start, end] of ranges) {
      let into = false;
      for (let index = start; index <= end; index++) {
        const gndAlt = altitude.altitudes[index];
        const alt = track.alt[index];
        const top = props.top + (props.flags & Flags.TopRefGnd ? gndAlt : 0);
        const bottom = props.bottom + (props.flags & Flags.FloorRefGnd ? gndAlt : 0);
        if (alt >= bottom && alt <= top) {
          into = true;
          break;
        }
      }
      intoRanges.push(into);
    }
  }

  // Build the proto object
  const aspObjects: any[] = [];
  for (const [key, ranges] of rangesByAirspaceKey) {
    const props = featureByKey.get(key)?.properties;
    const intoRanges = intoRangesByAirspaceKey.get(key) as boolean[];
    ranges.forEach(([start, end], rangeIndex) => {
      aspObjects.push({
        startSec: track.timeSec[start],
        endSec: track.timeSec[end],
        name: props.name,
        category: props.category,
        top: props.top,
        bottom: props.bottom,
        flags: props.flags,
        into: intoRanges[rangeIndex],
      });
    });
  }

  aspObjects.sort((a, b) => (a.startSec < b.startSec ? -1 : 1));

  const pAsp = protos.Airspaces.create({ hasErrors: altitude.hasErrors });

  aspObjects.forEach((asp) => {
    pAsp.startSec.push(asp.startSec);
    pAsp.endSec.push(asp.endSec);
    pAsp.name.push(asp.name);
    pAsp.category.push(asp.category);
    pAsp.top.push(asp.top);
    pAsp.bottom.push(asp.bottom);
    pAsp.flags.push(asp.flags);
    pAsp.into.push(asp.into);
  });

  return pAsp;
}

// Returns a unique if for an airspace feature.
function getAspFeatureKey(feature: any): string {
  const p = feature.properties;
  return `${p.name}-${p.bottom}-${p.top}`;
}

// Returns a lazily instantiated LRU for airspace buffers.
function getAspCache(): LRU<ArrayBuffer | null> {
  if (aspCache == null) {
    console.log(`ASP LRU Capacity = ${ASLP_LRU_CAPACITY}`);
    aspCache = lru<Buffer | null>(ASLP_LRU_CAPACITY);
  }
  return aspCache;
}
