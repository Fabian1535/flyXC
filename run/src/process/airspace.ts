import async from 'async';
import * as protos from 'flyxc/common/protos/track';
import { Flags, getAspTileUrl, isInFeature } from 'flyxc/common/src/airspaces';
import { pixelCoordinates } from 'flyxc/common/src/proj';
import { Point } from 'flyxc/common/src/runtime-track';
import { VectorTile } from 'mapbox-vector-tile';
import { LRU, lru } from 'tiny-lru';

import { httpsGet } from './request';

// Zoom level for the airspaces tiles.
const ZOOM_LEVEL = 12;

// Look MARGIN_METER above max altitude.
const MARGIN_METER = 200;

// Average airspace tile size.
// The average size is ~250 bytes at level 12 (395MB	for 1649364 tiles).
// The maximum size is ~2.2KB.
const DEFAULT_ASP_LRU_SIZE_MB = 80;
const ASP_SIZE_MB = 1000 / (1000 * 1000);

// null is used when there is no tile at the given location.
let aspCache: LRU<Buffer | null> | null = null;

// Retrieves the airspaces for the given track.
export async function fetchAirspaces(track: protos.Track, altitude: protos.GroundAltitude): Promise<protos.Airspaces> {
  // Retrieve the list of urls to download.
  const indexesByTileUrl = new Map<string, number[]>();
  const tilePixels: Array<Point> = [];

  track.lat.forEach((lat: number, i: number) => {
    const lon = track.lon[i];
    const { tile, px } = pixelCoordinates({ lat, lon }, ZOOM_LEVEL);
    tilePixels.push(px);
    const url = getAspTileUrl(tile.x, tile.y, ZOOM_LEVEL);
    if (!indexesByTileUrl.has(url)) {
      indexesByTileUrl.set(url, []);
    }
    indexesByTileUrl.get(url)?.push(i);
  });

  // Download the tiles.
  const bufferByTileUrl = new Map<string, Buffer | null>();
  await async.eachLimit(indexesByTileUrl.keys(), 5, async (url: string) => {
    let buffer = getAspCache().get(url);
    if (buffer === undefined) {
      try {
        buffer = await httpsGet(url);
      } catch (e) {
        // Downloading tiles might return 404 if there is not airspace.
        // We then can not rely on 404 to detect errors.
        buffer = null;
      }
    }
    getAspCache().set(url, buffer);
    bufferByTileUrl.set(url, buffer);
  });

  // Retrieve fix indexes for every airspaces.
  const gndAlt = altitude.altitudes;
  const indexesByAirspaceKey = new Map<string, number[]>();
  const featureByKey = new Map<string, any>();
  const maxAltitude = Math.max(...track.alt) + MARGIN_METER;
  for (const [url, indexes] of indexesByTileUrl) {
    const buffer = bufferByTileUrl.get(url);
    if (buffer == null) {
      continue;
    }
    const aspLayer = new VectorTile(new Uint8Array(buffer)).layers.asp;

    for (let i = 0; i < aspLayer.length; i++) {
      const feature = aspLayer.feature(i);
      const props = feature.properties;
      const key = getAspFeatureKey(feature);
      for (const fixIdx of indexes) {
        const flags = props.flags as number;
        const bottom = (props.bottom as number) + (flags & Flags.FloorRefGnd ? (gndAlt[fixIdx] as number) : 0);
        if (bottom > maxAltitude) {
          continue;
        }
        if (isInFeature(tilePixels[fixIdx], feature)) {
          if (!indexesByAirspaceKey.has(key)) {
            indexesByAirspaceKey.set(key, []);
            featureByKey.set(key, feature);
          }
          indexesByAirspaceKey.get(key)?.push(fixIdx);
        }
      }
    }
  }

  // Compute index ranges.
  const rangesByAirspaceKey = new Map<string, Array<[number, number]>>();
  for (const [key, indexes] of indexesByAirspaceKey) {
    indexes.sort((a, b) => (a < b ? -1 : 1));
    let start = indexes[0];
    let current = indexes[0];
    rangesByAirspaceKey.set(key, []);
    for (let i = 1; i < indexes.length; i++) {
      const next = indexes[i];
      if (next == current + 1) {
        current = next;
        continue;
      }
      rangesByAirspaceKey.get(key)?.push([start, current]);
      start = next;
      current = next;
    }
    rangesByAirspaceKey.get(key)?.push([start, current]);
  }

  // Check if the tracks crosses the airspaces.
  const intoRangesByAirspaceKey = new Map<string, boolean[]>();
  for (const [key, ranges] of rangesByAirspaceKey) {
    const props = featureByKey.get(key).properties;
    const intoRanges: boolean[] = [];
    intoRangesByAirspaceKey.set(key, intoRanges);
    for (const [start, end] of ranges) {
      let into = false;
      for (let index = start; index <= end; index++) {
        const gndAlt = altitude.altitudes[index];
        const alt = track.alt[index];
        const top = props.top + (props.flags & Flags.TopRefGnd ? gndAlt : 0);
        const bottom = props.bottom + (props.flags & Flags.FloorRefGnd ? gndAlt : 0);
        if (alt >= bottom && alt <= top) {
          into = true;
          break;
        }
      }
      intoRanges.push(into);
    }
  }

  // Build the proto object
  const aspObjects: any[] = [];
  for (const [key, ranges] of rangesByAirspaceKey) {
    const props = featureByKey.get(key)?.properties;
    const intoRanges = intoRangesByAirspaceKey.get(key) as boolean[];
    ranges.forEach(([start, end], rangeIndex) => {
      aspObjects.push({
        startSec: track.timeSec[start],
        endSec: track.timeSec[end],
        name: props.name,
        category: props.category,
        top: props.top,
        bottom: props.bottom,
        flags: props.flags,
        into: intoRanges[rangeIndex],
      });
    });
  }

  aspObjects.sort((a, b) => (a.startSec < b.startSec ? -1 : 1));

  const protoAirspaces = protos.Airspaces.create({ hasErrors: altitude.hasErrors });

  aspObjects.forEach((asp) => {
    protoAirspaces.startSec.push(asp.startSec);
    protoAirspaces.endSec.push(asp.endSec);
    protoAirspaces.name.push(asp.name);
    protoAirspaces.category.push(asp.category);
    protoAirspaces.top.push(asp.top);
    protoAirspaces.bottom.push(asp.bottom);
    protoAirspaces.flags.push(asp.flags);
    protoAirspaces.into.push(asp.into);
  });

  return protoAirspaces;
}

// Returns a unique if for an airspace feature.
function getAspFeatureKey(feature: any): string {
  const p = feature.properties;
  return p.name + '-' + p.bottom + '-' + p.top;
}

// Returns a lazily instantiated LRU for airspace buffers.
// Use the `ASP_LRU_SIZE_MB` environment variable to override the capacity.
function getAspCache(): LRU<Buffer | null> {
  if (aspCache == null) {
    const mb = Number(process.env.ASP_LRU_SIZE_MB || DEFAULT_ASP_LRU_SIZE_MB);
    const capacity = Math.floor(mb / ASP_SIZE_MB);
    console.log(`ASP LRU Capacity = ${capacity} - ${Math.round(capacity * ASP_SIZE_MB)}MB`);
    aspCache = lru<Buffer | null>(capacity);
  }
  return aspCache;
}
