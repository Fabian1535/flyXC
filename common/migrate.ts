/* eslint-enable @typescript-eslint/no-var-requires */
const { Datastore } = require('@google-cloud/datastore');
/* eslint-disable @typescript-eslint/no-var-requires */

import async from 'async';
import Pbf from 'pbf';

import { fetchAirspaces } from '../run/src/process/airspace';
import { retrieveTrackById, saveTrack, TrackEntity } from './datastore';
import {
  diffDecodeArray,
  diffDecodeTrack,
  diffEncodeAirspaces,
  ProtoAirspaces,
  ProtoGroundAltitude,
  ProtoGroundAltitudeGroup,
  ProtoTrack,
  ProtoTrackGroup,
} from './track';
import * as protos from './track_proto.js';

const BATCH_SIZE = 300;

async function addAirspaces(trackEntity: TrackEntity) {
  let migrated = false;
  const trackGroupBin = trackEntity.track_group;
  const gndAltGroupBin = trackEntity.ground_altitude_group;
  const aspGroupBin = trackEntity.airspaces_group;

  if (trackGroupBin && gndAltGroupBin && !aspGroupBin) {
    // Retrieve the tracks.
    const trackGroup: ProtoTrackGroup = (protos.TrackGroup as any).read(new Pbf(trackGroupBin));
    const tracks: ProtoTrack[] = trackGroup.tracks.map(diffDecodeTrack);

    const gndAltGroup: ProtoGroundAltitudeGroup = (protos.GroundAltitudeGroup as any).read(new Pbf(gndAltGroupBin));
    const gndAlts: ProtoGroundAltitude[] = gndAltGroup.ground_altitudes.map((g) => ({
      altitudes: diffDecodeArray(g.altitudes, 1),
      has_errors: g.has_errors,
    }));

    // Add airspaces.
    const airspaces: ProtoAirspaces[] = [];
    await async.eachOfSeries(tracks, async (track, i) => {
      airspaces.push(await fetchAirspaces(track, gndAlts[Number(i)]));
    });
    const pbf = new Pbf();
    (protos.AirspacesGroup as any).write({ airspaces: airspaces.map(diffEncodeAirspaces) }, pbf);
    trackEntity.airspaces_group = Buffer.from(pbf.finish());
    migrated = true;
  }

  if ((trackEntity as any).data) {
    delete (trackEntity as any).data;
    migrated = true;
  }

  return migrated;
}

// Add airspaces to the tracks.
export async function migrate(): Promise<void> {
  console.log('Migrate');
  const datastore = new Datastore();

  let token: string | undefined;

  let query = datastore.createQuery('Track').select('__key__').limit(BATCH_SIZE);
  let numMigrated = 0;
  let numBatches = 0;

  while (true) {
    console.log(`Processing batch #${numBatches} (${numBatches * BATCH_SIZE} tracks).`);
    numBatches++;
    if (token != null) {
      query = query.start(token);
    }
    const [entities, info] = await datastore.runQuery(query);
    if (entities.length == 0 || info.moreResults == Datastore.NO_MORE_RESULTS) {
      break;
    }
    token = info.endCursor;
    await async.eachLimit(entities, 10, async (entity: any) => {
      const id = entity[datastore.KEY].id;
      const trackEntity = await retrieveTrackById(id);
      if (trackEntity) {
        console.log(`migrating id: ${id} hash: ${trackEntity.hash} (${numMigrated} tracks migrated)`);
        const migrated = await addAirspaces(trackEntity);
        if (migrated) {
          numMigrated++;
          await saveTrack(trackEntity);
        }
      }
    });
  }
}
