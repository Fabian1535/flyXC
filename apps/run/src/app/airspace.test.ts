import { diffDecodeTrack, protos } from '@flyxc/common';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fetchAirspaces, getIndicesByUrlAndPx } from './airspace';

function readTrack(filename: string): protos.Track {
  const pbf = readFileSync(path.join(__dirname, 'fixtures', filename));
  const { metaTrackGroups } = protos.MetaTrackGroupsAndRoute.fromBinary(new Uint8Array(pbf));

  expect(metaTrackGroups).toHaveLength(1);
  expect(metaTrackGroups[0].trackGroupBin).not.toBeNull();
  const trackGroup = protos.TrackGroup.fromBinary(metaTrackGroups[0].trackGroupBin);

  expect(trackGroup.tracks).toHaveLength(1);
  return diffDecodeTrack(trackGroup.tracks[0]);
}

describe('Airspaces', () => {
  describe('Extract info from tracks', () => {
    it('Number of tiles for a big track', () => {
      const track = readTrack('570kms.pbf');
      const { indexesByTileUrl, pxInTiles } = getIndicesByUrlAndPx(track);
      expect(indexesByTileUrl.size).toEqual(74);
      expect(pxInTiles).toHaveLength(4915);
      expect(indexesByTileUrl.has('https://airsp.storage.googleapis.com/tiles/13/3284/4244.pbf'));
    });

    it('Number of tiles for a huge track', () => {
      const track = readTrack('huge.pbf');
      const { indexesByTileUrl, pxInTiles } = getIndicesByUrlAndPx(track);
      expect(indexesByTileUrl.size).toEqual(474);
      expect(pxInTiles).toHaveLength(6236);
    });

    it('Should support a upper bound of tiles', () => {
      const track = readTrack('570kms.pbf');
      const { indexesByTileUrl, pxInTiles } = getIndicesByUrlAndPx(track, 5);
      expect(indexesByTileUrl.size).toEqual(5);
      expect(pxInTiles).toHaveLength(523);
      expect(indexesByTileUrl.has('https://airsp.storage.googleapis.com/tiles/13/3284/4244.pbf'));
    });
  });

  describe('fetchAirspaces', () => {
    it('should find the airspaces', async () => {
      const track = readTrack('570kms.pbf');
      const gndAltitude = protos.GroundAltitude.create({
        altitudes: new Array(track.lat.length).fill(0),
        hasErrors: false,
      });
      expect(await fetchAirspaces(track, gndAltitude)).toMatchSnapshot();
    });
  });
});
