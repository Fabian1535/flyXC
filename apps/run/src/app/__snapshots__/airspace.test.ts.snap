// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airspaces fetchAirspaces should find the airspaces 1`] = `
Object {
  "bottom": Array [
    1067,
    0,
    1829,
    0,
  ],
  "category": Array [
    "C",
    "F",
    "F",
    "G",
  ],
  "endSec": Array [
    1666952265,
    1666956673,
    1666956657,
    1666979201,
  ],
  "flags": Array [
    1,
    18,
    2,
    34,
  ],
  "hasErrors": false,
  "into": Array [
    false,
    true,
    false,
    false,
  ],
  "name": Array [
    "TMA-NATAL",
    "SBR223- WHISKY_DOIS",
    "SBR224- WISKY100",
    "G - SBR266-QUIXADA",
  ],
  "startSec": Array [
    1666949217,
    1666952377,
    1666952377,
    1666978953,
  ],
  "top": Array [
    4420,
    8839,
    4572,
    1006,
  ],
}
`;
